### ISSUE TITLE
Feature Request: Support indices raising and lowering in N-dimensional numpy arrays

### ISSUE THREAD

**Issue by ritzvik** [numpy/numpy#13965]
Indices raising and lowering are a very crucial part of differential geometry and particularly important in relativistic physics.

See https://en.wikipedia.org/wiki/Raising_and_lowering_indices

Example : 
  - A 4D tensor can have index configuration `ulll`, and can be changed to `lulu` with the help of a metric tensor, here `u` denotes upper index and `l` denotes lower index.

### Reproducing code example:

```python
import numpy as np
T = np.zeros(shape=(4,4,4,4), dtype=float)
# substitute some values in T
M = np.zeros(shape=(4,4), stype=float)
# substitute some values in M(Metric Tensor)
T_ = np.change_config(T, metric=M, old='ulll', new='lulu')
print(T_)
# Example snippet
```

**Comment by eric-wieser** [comment]
If I understand that function, the implementation is:
```python
def change_config(arr, *, metric, old, new):
   for axis, (o, n) in enumerate(zip(old, new)):
       if o != n:
           arr = np.moveaxis(np.moveaxis(arr, i, -1) @ metric), -1, i)
    return arr
```

You might find it's better to make an `ndarray` subclass that keeps track of the lower/upper-ness of the axes

**Comment by ritzvik** [comment]
Yes `moveaxis` is required, and it also requires `tensorproduct` and `tensorcontraction` which is provided `einsum` in `np`, I guess. I couldn't understand the `@ metric` thing. Sorry for the trouble!

See https://github.com/einsteinpy/einsteinpy/blob/master/src/einsteinpy/symbolic/tensor.py#L29-L63 . And because of change to any arbitrary configuration, it may require a series of `tensorproduct`s and contractions.

**Comment by eric-wieser** [comment]
`@` here is matrix multiplication, which I believe subsumes contraction and product for the cases that matter here - assuming that a metric tensor is always 2D, which your links seem to suggest is so.

I'm not sure this is a good fit for numpy - on the one hand, we don't really have a true concept of tensors - but on the other, we do provide `einsum` and `tensordot`...

**Comment by ritzvik** [comment]
> `@` here is matrix multiplication, which I believe subsumes contraction and product for the cases that matter here - assuming that a metric tensor is always 2D, which your links seem to suggest is so.
> 
> I'm not sure this is a good fit for numpy - on the one hand, we don't really have a true concept of tensors - but on the other, we do provide `einsum` and `tensordot`...

`@` or `matmul` is not suitable for this as sum over any arbitrary index may be required. I saw that `tensordot` does the work of both `tensorproduct` and `contraction` combined, but still we need to permute the indices after doing `tensordot` using `moveaxis` that too a number of times to take into account the multiple number of raising and lowering that could be done in one change. I was saying that numpy has all the groundwork to implement this, I can take this up in PR.

Thanks for the quick reply 😃 

Should I go ahead with a PR?
